src/lib/badge/badge.ts

```diff
@@ -0,0 +1,162 @@
+    '[class.mat-badge-small]': 'matBadgeSize === "small"',
+    '[class.mat-badge-medium]': 'matBadgeSize === "medium"',
+    '[class.mat-badge-large]': 'matBadgeSize === "large"',
+    '[class.mat-badge-hidden]': 'matBadgeHidden',
+  },
+})
+export class MatBadge {
+
+  /** The color of the badge. Can be `primary`, `accent`, or `warn`. */
+  @Input('matBadgeColor')
+  get color(): ThemePalette { return this._color; }
+  set color(value: ThemePalette) {
+    this._color = value;
+    this._setColor(value);
+  }
+  private _color: ThemePalette = 'primary';
+
+  /** Whether the badge should overlap its contents or not */
+  @Input('matBadgeOverlap')
+  get overlap(): boolean { return this._overlap; }
+  set overlap(val: boolean) {
+    this._overlap = coerceBooleanProperty(val);
+  }
+  private _overlap: boolean = true;
+
+  /**
+   * Position the badge should reside.
+   * Accepts any combination of 'above'|'below' and 'before'|'after'
+   */
+  @Input('matBadgePosition')
+  get position(): string { return this._position; }
+  set position(val: string) {
+    this._position = val;
+    this._isAbove = val.indexOf('below') === -1;
+    this._isAfter = val.indexOf('before') === -1;
+  }
+  private _position: string = 'above after';
+
+  @Input('matBadge')
+  get content(): string { return this._content; }
+  set content(val: string) {
+    this._content = val;
+    this._updateTextContent();
+  }
+  private _content: string;
+
+  /** Message used to describe the decorated element via aria-describedby */
+  @Input('matBadgePosition')
+  get description(): string { return this._description; }
+  set description(val: string) {
+    this._setLabel(val, this._description);
+    this._description = val;
+  }
+  private _description: string;
+
+  /** Size of the badge. 'small' | 'medium' | 'large' */
+  @Input() matBadgeSize: string = 'medium';
+
+  /** Toggle the visibility of the badge on the host element. */
+  @Input()
+  set matBadgeHidden(val: boolean) {
+    this._hidden = coerceBooleanProperty(val);
+  }
+  get matBadgeHidden(): boolean {
+    return this._hidden;
+  }
+  private _hidden: boolean;
+
+  /** Unique id for the badge */
+  _id: number = nextId++;
+
+  /** Whether the badge is above the host or not */
+  _isAbove: boolean = true;
+
+  /** Whether the badge is after the host or not */
+  _isAfter: boolean = true;
+
+  constructor(
+      private _renderer: Renderer2,
+      private _elementRef: ElementRef,
+      private _ariaDescriber: AriaDescriber) {}
+
+  /** Injects a span element into the DOM with the content. */
+  private _updateTextContent(): HTMLSpanElement {
+    let content = document.getElementById(`mat-badge-content-${this._id}`);
+
+    if (!content) {
+      content = this._renderer.createElement('span');
+
+      content = document.createElement('span');
+      content.setAttribute('id', `mat-badge-content-${this._id}`);
+      content.classList.add('mat-badge-content');
+      content.textContent = this.content;
```

Comment:
Thought about this a bit more: I think we should create something like a `MatBadgeContent` component and then a `DomPortalHost` to put it in place. This would let us associate the non-color/typography styles with that component instead of having to put it all in the theme (this is what we do with tooltip except it wouldn't be overlay-based here).